### DISCUSSION TITLE
Is there any way to convert skywalking metrics into prometheus?

### DISCUSSION THREAD

**Discussion by ravenfu** [apache/skywalking#7504] [Q&A] [answered]
Please answer these questions before submitting your issue.

- Why do you submit this issue?
- [x] Question or discussion
- [ ] Bug
- [ ] Requirement
- [ ] Feature or performance improvement

___
Is there any way to convert skywalking metrics into prometheus?

I'm designing our monitor system. I'm planning to monitor applications by using skywalking, monitor infrastructure and middleware by using prometheus, collect and analysis logs by using Loki, and show the system state on Grafana.

Skywalking is the best tool on tracing, but I want to unified alarm on Prometheus, and unified display on Grafana for Monitoring Screen.

So I want to transfer skywalking metrics to prometheus, not only it's self monitoring metrics.
Is this idea feasible?

**Comment by wu-sheng** [comment]
It depends on what metrics are you going to do. If you want to forward all, Prometheus would crash, it has no chance to support SkyWalking's whole metrics dataset.

**Reply by ravenfu** [reply]
We most care about:
1. jvm status of each instance, metrics series of instance_jvm_*
2. only entrypoint metrics of each service, metrics series of endpoint_*. special service like gateway or nginx

**Reply by wu-sheng** [reply]
Then, it is unrealistic. The number of endpoints is easier over 10 million. Prometheus has no chance to survive, which is also why we didn't build SkyWalking like they did.

**Reply by ravenfu** [reply]
1. I tried to set the Grafana datasource to Elasticsearch, but the format of time field, `time_bucket `,  is Long, Grafana not support. OMG
2. I'm afraid the single point of failure if i write a converter to read from ES and then push to Pushgateway. And your worry of data size.

So it seems impossible.

**Reply by wu-sheng** [reply]
> I tried to set the Grafana datasource to Elasticsearch, but the format of time field, time_bucket , is Long, Grafana not support. OMG

We know that for long. And no one is willing to dig deeper to see whether it is meaningful if SkyWalking put the `time_bucket` as a timestamp, and how it could work in Grafana. 
Until some one works on that, and open source the solution one day. I think I don't know exactly what this could be.

> I'm afraid the single point of failure if i write a converter to read from ES and then push to Pushgateway. And your worry of data size.

If you are running for endpoint, you must fail. And pushgateway doc gives a very clear introduction, it is not for this kind of usage.

**Reply by nisiyong** [reply]
> We know that for long. And no one is willing to dig deeper to see whether it is meaningful if SkyWalking put the `time_bucket` as a timestamp, and how it could work in Grafana.
> Until some one works on that, and open source the solution one day. I think I don't know exactly what this could be.

We want to show skywalking metrics in Grafana, too. And today I took some effort to adapt that.

Cause the Grafana Elasticsearch datasource only supports the date type at the time field, I create a new index mapping for `service_cpm` to test it. The following mappings change the `time_bucket` type as date type:
```json
{
    "mappings": {
        "type": {
            "properties": {
                "entity_id": {
                    "type": "keyword"
                },
                "time_bucket": {
                    "type": "date",
                    "format": "yyyyMMdd||yyyyMMddHH||yyyyMMddHHmm"
                },
                "total": {
                    "type": "long",
                    "index": false
                },
                "value": {
                    "type": "long"
                }
            }
        }
    }
}
```

And then I [reindex](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html) the `service_cpm-20220323` data to `service_cpm` from the following scripts: 
```bash
curl -X POST  -H 'Content-Type: application/json' -d '{
    "source": {
        "index": " service_cpm-20220323"
    },
    "dest": {
        "index": "service_cpm"
    }
}' http://$es/_reindex?pretty
```
Now the datasource could be setup, and I could show the data in Grafana.
<img width="452" alt="image" src="https://user-images.githubusercontent.com/8198862/159672702-4cd4d4c6-6183-4feb-b240-86c6a9dd81a8.png">
<img width="1446" alt="image" src="https://user-images.githubusercontent.com/8198862/159672972-21bf815b-393b-410d-a293-c73b5d227a4e.png">

But there have some problems: 

1. The `time_bucket` filed is not timestamped value, it has the time-zone problem after I reindex the data.
2. The `time_bucket` field contains the downsampling data, like hourly metrics and daily metrics, so if we want to fix the 1st problem, we may need to add a new field to distinguish the metrics.
3. The `entity_id` filed needs to decode, if stored a new filed `entity_name` will be friendly for users when they show data in Grafana.

So if we want to show the skywalking Elasticsearch data in Grafana, we may do the following things:
1. Add a new date field like `timestamp` to replace`time_bucket`
2. Add a new field  to distinguish the metrics from minute, hour, or day
3. Add a new field to store name, like `entity_name`

Grafana is a good tool to show observer data, I wonder how the community considers it. I think we could discuss #2053 again.

**Reply by wu-sheng** [reply]
entity name is out of box, I think we have the configuration already.

The key point is the numbers of candidates. I have seens hundreds and thousands of services and pods per service, and countless(100k+) endpoints. How does this work in grafana as labels in graphs.

**Reply by nisiyong** [reply]
Well, I agree with your point. But I think how to show graphs is another problem, maybe users could use the select options component to filter the data, or some scenarios only show the TopN. Just like skywalking-rocketbot-ui does.

I wanner said, how to show graphs efficiently is the user-side problem, and whether skywalking is possible to show data in Grafana is another problem.

**Reply by wu-sheng** [reply]
Once we said, SkyWalking officially could do, it is our problem.
This is a difference when you share a blog about practice V.S. official doc says. 😂

**Reply by ravenfu** [reply]
Grafana ecosystem now has 'tempo' and 'loki', those products can do tracing and log analysis.
When we build a monitor platform,  operations staff should to know what happened first , that is Grafana ecosystem is good at. And then developers use skywalking to find out where the system was wrong.
So I think this issue's core problem is we need skywallking at level 1 to show the problem of the system, the event, the metrics, although we known the opentracing design about endpoint is very helpful

**Reply by wu-sheng** [reply]
@ravenfu Don't get me wrong, I am not saying SkyWalking and Prometheus should pick one, we are designed for different scenarios. Users want to use both are fine.
 
My point is, none of SkyWalking's PMC is an expert on Prometheus, and we don't use Prometheus in the company or product to cover SkyWalking's scope. So we don't have enough knowledge that how to make SkyWalking work well with Grafana.

If someone really wants SkyWalking to move on this integration seriously, we need them to step out and work with the community to resolve above real issues. We, as a community has been widely used, can't say `we have a half-completed feature, you could try and suffer, and no one in the PMC team even used.`. That is much worth than saying `we don't support that`.

**Reply by nisiyong** [reply]
Maybe I could have a try, let's see what happens.

**Reply by wu-sheng** [reply]
@nisiyong Thanks. once you have a proposal, please start a new discussion. This one is old, and it is about converting to Prometheus, I think what you did is showing on Grafana.

**Reply by liyuntian** [reply]
> Maybe I could have a try, let's see what happens.

then what happen？we want to collect skywalking metrics in different k8s colony in one place

**Reply by wu-sheng** [reply]
Collecting SkyWalking metrics or service metrics in SkyWalking are different. Define what do you propose before doing anything.

**Reply by liyuntian** [reply]
we want to collect service metrics and Jvm metrics in different SkyWalking,and display in our monitor system，so we can monitor multiple cluster services and jvm in one interface

**Reply by nisiyong** [reply]
@liyuntian If you want to do this, you could do it yourself. Some tips for you, you could collect the jvm metrics which foward by grpc gateway, and store it on your own application.

**Reply by wu-sheng** [reply]
> we want to collect service metrics and Jvm metrics in different SkyWalking,and display in our monitor system，so we can monitor multiple cluster services and jvm in one interface

We don't support this directly, you could use metrics exporter to forward metrics on your own.

**Comment by wu-sheng** [comment, marked as answer]
This is what we are going to do officially.

https://github.com/apache/skywalking/issues/10080

PromQL based query support could support Grafana and any other 3rd party integrated with Prometheus, as they are sharing to use this query protocol.
